br/pkg/restore/snap_client/tikv_sender.go

```diff
@@ -119,16 +179,103 @@ func SortAndValidateFileRanges(
 			zap.Int("Merged(keys avg)", stat.MergedRegionKeysAvg),
 			zap.Int("Merged(bytes avg)", stat.MergedRegionBytesAvg))
 
+		// skip some ranges if recorded by checkpoint
+		// Notice that skip ranges after select split keys in order to make the split keys
+		// always the same.
+		checkpointSet := checkpointSetWithTableID[table.NewPhysicalID]
+
+		// Generate the split keys, and notice that the way to generate split keys must be deterministic
+		// and regardless of the current cluster region distribution. Therefore, when restore fails, the
+		// generated split keys keep the same as before the next time we retry to restore.
+		//
+		// Here suppose that all the ranges is in the one region at beginning.
+		// In general, the ids of tables, which are created in the previous stage, are continuously because:
+		//
+		// 1. Before create tables, the cluster global id is allocated to ${GLOBAL_ID};
+		// 2. Suppose the ids of tables to be created are {t_i}, which t_i < t_j if i < j.
+		// 3. BR preallocate the global id from ${GLOBAL_ID} to t_max, so the table ids, which are larger
+		//  than ${GLOBAL_ID}, has the same downstream ids.
+		// 4. Then BR creates tables, and the table ids, which are less than or equal to ${GLOBAL_ID}, are
+		//  allocated to [t_max + 1, ...) in the downstream cluster.
+		// 5. Therefore, the BR-created tables are usually continuously.
+		//
+		// Besides, the prefix of the existing region's start key and end key should not be `t{restored_table_id}`.
 		for _, rg := range sortedRanges {
-			sortedSplitKeys = append(sortedSplitKeys, rg.EndKey)
+			// split key generation
+			afterMergedGroupSize := groupSize + rg.Size
+			afterMergedGroupCount := groupCount + rg.Count
+			if afterMergedGroupSize > splitSizeBytes || afterMergedGroupCount > splitKeyCount || mergedRangeCount > MergedRangeCountThreshold {
+				log.Info("merge ranges across tables due to kv size/count or merged count threshold exceeded",
+					zap.Uint64("merged kv size", groupSize),
+					zap.Uint64("merged kv count", groupCount),
+					zap.Int("merged range count", mergedRangeCount))
+				groupSize, groupCount = rg.Size, rg.Count
+				mergedRangeCount = 0
```

Comment:
shall we reset this value after `lastFilesGroup` appended?

Reply:
`mergedRangeCount` records the number of files in `tableIDWithFilesGroup` to make sure the batch size of ingest is not too large.